### ISSUE TITLE
 SGD, RMSprop, Adam에 대해서 아는대로 설명한다면?

### ISSUE THREAD

**Issue by kimyuuuuuu** [AInsteins/AI-Tech-Interview#50]

**Comment by kimyuuuuuu** [comment]
## SGD

- Batch Gradient Descent 으로 계산시 한번 step을 내딛을 때 전체 데이터에 대해 Loss Function을 계산해야 하므로 **너무 많은 계산량이 필요**하다. 
• 전체 훈련 데이터셋을 대상으로 학습하는 것은 한정된 리소스를 가지고 있는 우리의 분석 환경에서 매우 비효율적이며, 파라미터 업데이트 수가 적다는 것은 랜덤 하게 뽑힌 시작 위치의 가중치 수도 적으므로, Local minimum 현상이 발생할 확률도 높다는 것이다.
→ 이를 **방지**하기 위해 **SGD(Stochastic Gradient Descent, 확률적 경사 하강법)** 사용
    - Batch Gradient Descent
        - **전체 데이터 셋**에 대한 에러를 구한 뒤 기울기를 한번만 계산하여 모델의 parameter 를 업데이트
        - 경사 하강법 대상이 배치 크기와 동일하다는 것이다.
        - 데이터셋 모두를 대상으로 하다 보니 파라미터가 한번 이동할 때마다, 계산해야 할 값이 지나치게 많으므로, **계산 시간도 엄청 길어지고, 소모되는 메모리도 엄청**나다.
        - mini batch 안 모든 데이터를 대상으로 경사 하강법을 실시하므로, **안정적**으로 수렴한다. → 수렴까지 발생하는 총 파라미터 업데이트 수는 매우 적다.
        - 안정적으로 수렴하는 것은 좋으나, 안정적으로 움직이기 때문에 지역 최소해(Local Minimum)에 빠지더라도 안정적으로 움직이므로 빠져나오기 힘들다. **즉, Local Optima(minimum) 문제가 발생할 가능성이 높다.**
        - **학습 데이터셋이 커지면 커질수록 시간과 리소스 소모가 지나치게 크다.**
        
- SGD
    - loss function을 계산할 때 전체 데이터(batch) 대신 데이터 한 개 또는 일부 조그마한 데이터의 모음**(mini-batch)에 대해서만 loss function을 계산**한다.
    = 추출된 데이터 **한 개**에 대해서 error gradient 를 계산하고, Gradient descent 알고리즘을 적용
    - **학습 데이터셋에서 무작위로 한 개의 샘플 데이터 셋을 추출하고, 그 샘플에 대해서만 기울기를 계산**
    - **샘플 데이터 셋에 대해서만 경사(Gradient)를 계산**하므로, 매 반복에서 다뤄야 할 데이터 수가 매우 적어, **학습 속도가 매우 빠르다.**
    - 데이터 한 개 사용 : Stochastic Gradient Descent(SGD)
    데이터의 일부(mini-batch) 사용: mini-batch Stochastic Gradient Descent(mini-batch SGD)
    → 오늘날 일반적으로 통용되는 SGD는 **mini-batch SGD**이다.
    - 이 방법은 batch gradient descent 보다 다소 부정확할 수는 있지만, 훨씬 계산 속도가 빠르기 때문에 같은 시간에 더 많은 step을 갈 수 있으며 여러 번 반복할 경우 보통 batch의 결과와 유사한 결과로 수렴한다.
    - 손실 함수가 최솟값에 다다를 때까지 위아래로 요동치며 움직이다 보니, 학습이 진행되다 보면, 최적해에 매우 근접하게 움직이긴 하겠으나, **최적해(Global minimum)에 정확히 도달하지 못할 가능성이 있다.**
    - 그러나, 이렇게 요동치며 움직이므로, **지역 최솟값(Local minimum)에 빠진 다할지라도, 지역 최솟값에서 쉽게 빠져나올 수 있으며, 그로 인해 전역 최솟값(Global minimum)을 찾을 가능성이 BGD에 비해 더 높다.**
    - 즉, **확률적 경사 하강법(SGD)은 속도가 매우 빠르고 메모리를 적게 먹는다는 장점이 있으나, 경사를 구할 때, 무작위성을 띄므로 지역 최솟값에서 탈출하기 쉬우나, 전역 최솟값에 다다르기 힘들다는 단점을 가지고 있다.**
    - SGD를 사용할 경우 Batch Gradient Descent에서 빠질 local minima에 빠지지 않고 더 좋은 방향으로 수렴할 가능성도 있다. 
    → 모델의 레이어 층은 하나의 행렬곱으로 생각할 수 있고, 여러개의 묶음 데이터는 행렬이라고 생각 할 수 있다.즉, 여러개의 묶음 데이터를 특정 레이어 층에 입력하는 것은 행렬 x 행렬로 이해할 수 있는데, SGD는 **입력 데이터 한 개만을 사용하기 때문에 한 개의 데이터를 '벡터' 로 표현하여 특정 레이어 층에 입력하는 것으로 이해할 수 있고 이는 벡터 x 행렬 연산이 된다.**
    →Shooting 이 일어나기 때문에 local optimal 에 빠질 리스크가 적다.


---

## RMSprop

- 딥러닝의 대가 제프리 힌톤이 제안한 방법으로서, **Adagrad**의 단점을 해결하기 위한 방법
- Adagrad의 식에서 gradient의 제곱값을 더해나가면서 구한 $G_t$부분을 합이 아니라 지수평균으로 바꾸어서 대체한 방법이다.
    

    
- 이렇게 대체를 할 경우 Adagrad처럼 $G_t$가 **무한정 커지지는 않으면서 최근 변화량의 변수간 상대적인 크기 차이는 유지**할 수 있다.

---

## Adam

- Adam(Adaptive Moment Estimation)은 RMSProp과 Momentum 방식을 합친 것 같은 알고리즘이다.
- 이 방식에서는 Momentum 방식과 유사하게 지금까지 계산해온 기울기의 지수평균을 저장하며, RMSProp과 유사하게 기울기의 제곱값의 지수평균을 저장한다.

**Comment by miinkang** [comment]
- Optimizer
    - loss function의 결과값을 최소화하는 모델의 파라미터를 찾는 것
- SGD (확률적 경사 하강법)
    - 한 스텝마다 loss function을 계산하면 계산량이 너무 많음
    - 이를 방지하기위한 방법
    - 전체 batch가 아닌 mini batch 단위로 loss function을 계산하는 방법
    - 여러번 반복하면 보통 전체 batch 결과와 유사한 결과로 수렴할 수 있음
    - local minima에 빠지지 않을 수 있으며, 계산 속도가 빠르다
- RMSprop
    - Adagrad의 단점을 해결하기 위한 방법
        - adagrad : 학습률 감소를 사용하는 방법
    - 기울기를 제곱해서 더하기 때문에 갱신되는 정도가 약하며 결국엔 갱신량이 0이 된다는 단점
    - 지수이동평균을 이용해 기울기 반영 정도를 다르게 하는 방법
- Adam
    - 학습률, 모멘텀용 계수 (1차, 2차)
    - 위 3개의 하이퍼파라미터를 설정해 사용한다.

**Comment by import-max** [comment]
### SGD

- 데이터 셋을 모두 사용하여 loss function 을 계산하는 걸 batch gradient descent라고 하는데, 너무 많은 데이터를 사용하는 것은 비효율적이라 하나의 데이터나 일부 데이터만 사용하는 등의 방법을 사용하는 것을 Stochastic Gradient Descent(SGD), mini-batch Stochastic Gradient Descent(mini-batch SGD)라고 한다.
- 장점은 local minima 에 빠지지 않을 수 있고, 효율적인 연산을 통해 global minima의 근사값에 도달이 가능하다는 점이다. 단점은, loss가 정확하지 않을 수 있다는 점이다.

### RMSprop

- Adagrad의 단점을 해결하기 위한 방법으로, 기울기를 단순 누적하지 않고 지수 가중 이동 평균 Exponentially weighted moving average 를 사용하여 최신 기울기들이 더 크게 반영되도록 하였다. 전 AdaGrad 의 h 에 새로운 hyper parameter ρ 를 추가하였다. 이제 h 가 무한히 커지지 않으면서 ρ 가 작을 수록 가장 최신의 기울기를 더 크게 반영하게 된다.

### Adam

- Momentum과 Adagrad를 융합한 방법으로, v, h 가 각각 최초 0으로 설정되어 학습 초반에 0으로 biased 되는 문제를 해결하기 위해 고안한 방법이다.

**Comment by jarammm** [comment]
1. sgd
    1. 계산할 때 전체 데이터(batch) 대신 데이터 1개나 mini-batch에 대한 loss function 계산
    2. 장점 : 빠른 속도. 정확한 최소 loss 지점을 찾기보다는 빨리 최솟값의 지점과 근사한 지점을 찾아내는 것에 중점을 둠
    3. 비등방성 함수(방향에 따라 성질, 즉 기울기가 달라지는 함수)에서의 탐색 경로가 비효율적임
2. rmsprop
    1. adagrad를 개선한 기법으로 과거의 모든 기울기를 균일하게 더하지 않고, 마치 LSTM처럼 먼 과거의 기울기는 서서히 잊고 새로운 기울기 정보를 크게 반영
    2. 장점 : 
        - 변수(feature)마다 적절한 학습률을 적용하여 효율적인 학습을 진행할 수 있다는 점
        - Adagrad보다 학습을 오래 할 수 있다는 점
            - adagrad는 학습 진행할수록 갱신 강도가 약해지면서 갱신량이 0이 되어 그 시점부터 전혀 갱신 안됨 → rmsprop은 지수이동평균의 업데이트 계수를 활용해 번째 time step의 누적 기울기(gt)가 무한정 커지는 것을 방지하여 오래 학습 가능
3. adam
    1. 이름은 adagrad + momentum이라는 뜻이고, RMSProp(adagrad 개선한 기법) + Momentum의 장점 결합
    2. 학습의 방향과 크기(=Learning rate)를 모두 개선, 표준처럼 많이 쓰이는 최적화 기법
    3. Momentum 방식과 유사하게 지금까지 계산해온 기울기의 지수평균을 저장하며, RMSProp과 유사하게 기울기의 제곱값의 지수평균을 저장한다.
    4. 하이퍼파라미터 3개 설정 : 학습률(default : 0.001), 1차 모멘텀용 계수(default : 0.9), 2차 모멘텀용 계수(default : 0.999)
    5. 장점 : 하이퍼 파라미터의 **편향 보정** 진행, 모멘텀과 비슷한 패턴의 움직임을 보이지만 학습의 갱신 강도를 적응적으로 조정하여 흔들림이 적음
    6. (m과 v가 처음에 0으로 초기화되어 있기 때문에 학습의 초반부에서는 $m_t$(1-1차모멘텀용계수로 나눠 갱신), $v_t$(1-2차모멘텀용계수로 나눠 갱신)가 0에 가깝게 bias 되어있을 것이라고 판단하여 이를 unbiased 하게 만들어주는 작업을 거친다.)
